cpp/include/cudf/detail/utilities/stream_pool.hpp

```diff
@@ -25,6 +25,62 @@
 
 namespace cudf::detail {
 
+class cuda_stream_pool {
+ public:
+  // matching type used in rmm::cuda_stream_pool::get_stream(stream_id)
+  using stream_id_type = std::size_t;
+
+  virtual ~cuda_stream_pool() = default;
+
+  /**
+   * @brief Get a `cuda_stream_view` of a stream in the pool.
+   *
+   * This function is thread safe with respect to other calls to the same function.
+   *
+   * @return Stream view.
+   */
+  virtual rmm::cuda_stream_view get_stream() = 0;
+
+  /**
+   * @brief Get a `cuda_stream_view` of the stream associated with `stream_id`.
+   *
+   * Equivalent values of `stream_id` return a `cuda_stream_view` to the same underlying stream.
+   * This function is thread safe with respect to other calls to the same function.
+   *
+   * @param stream_id Unique identifier for the desired stream
+   * @return Requested stream view.
+   */
+  virtual rmm::cuda_stream_view get_stream(stream_id_type stream_id) = 0;
+
+  /**
+   * @brief Get a set of `cuda_stream_view` objects from the pool.
+   *
+   * An attempt is made to ensure that the returned vector does not contain duplicate
+   * streams, but this cannot be guaranteed if `count` is greater than the value returned by
+   * `get_stream_pool_size()`.
+   *
+   * This function is thread safe with respect to other calls to the same function.
+   *
+   * @param count The number of stream views to return.
+   * @return Vector containing `count` stream views.
+   */
+  virtual std::vector<rmm::cuda_stream_view> get_streams(std::size_t count) = 0;
+
+  /**
+   * @brief Get the number of unique stream objects in the pool.
+   *
+   * This function is thread safe with respect to other calls to the same function.
+   *
+   * @return the number of stream objects in the pool
+   */
+  virtual std::size_t get_stream_pool_size() const = 0;
```

Comment:
The name is a bit redundant. We know that this is a stream pool.
```suggestion
  virtual std::size_t get_pool_size() const = 0;
```
or just `size()`.

Reply:
`get_pool_size()` is consistent with `rmm::cuda_stream_pool`. Surprised this wasn't caught before.

Reply:
yeah, should be `get_pool_size`. Not a change for this PR, IMO.

Reply:
Okay, but then please address it in a follow up work.

Reply:
Thank you for pointing this out! I'll work on this change in a follow-up PR.

Reply:
Why isn't this a change for this PR, if all the code in this section is new in this PR?